pallets/payment-streams/src/utils.rs

```diff
@@ -550,180 +552,389 @@ where
+                                total_amount_charged
+                                    .checked_add(&amount_to_charge)
+                                    .is_some(),
+                                ArithmeticError::Overflow
                             );
-                            payment_stream.last_charged_tick = last_chargeable_tick;
-                            Ok::<(), DispatchError>(())
-                        },
-                    )?;
 
-                    // Update the total amount charged:
-                    total_amount_charged = total_amount_charged
-                        .checked_add(&amount_to_charge)
-                        .ok_or(Error::<T>::ChargeOverflow)?;
+                            // Charge the payment stream from the user's balance
+                            T::NativeBalance::transfer(
+                                user_account,
+                                &provider_payment_account,
+                                amount_to_charge,
+                                Preservation::Preserve,
+                            )?;
+
+                            // Set the last charged price index to be the price index of the last chargeable tick
+                            DynamicRatePaymentStreams::<T>::mutate(
+                                provider_id,
+                                user_account,
+                                |payment_stream| {
+                                    let payment_stream = expect_or_err!(
+                                        payment_stream,
+                                        "Payment stream should exist if it was found before.",
+                                        Error::<T>::PaymentStreamNotFound
+                                    );
+                                    payment_stream.price_index_when_last_charged =
+                                        price_index_at_last_chargeable_tick;
+                                    Ok::<(), DispatchError>(())
+                                },
+                            )?;
+
+                            // Update the total amount charged:
+                            total_amount_charged = total_amount_charged
+                                .checked_add(&amount_to_charge)
+                                .ok_or(Error::<T>::ChargeOverflow)?;
+                        }
+                    }
                 }
             }
         }
 
-        // If the dynamic-rate payment stream exists:
-        if let Some(dynamic_rate_payment_stream) = dynamic_rate_payment_stream {
-            // Calculate the difference between the last charged price index and the price index at the last chargeable tick
-            // Note: If the last chargeable price index is less than the last charged price index, we charge 0 to the user, because that would be an impossible state.
+        Ok(total_amount_charged)
+    }
+
+    pub fn do_pay_outstanding_debt(user_account: &T::AccountId) -> DispatchResult {
```

Comment:
This is releasing the held funds from the user, right?

Reply:
What happens if the user's deposit is greater than what is owed to that Provider?

Reply:
Same as above!